hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractCompletableFuture.java

```diff
@@ -110,6 +167,32 @@ public V get() throws InterruptedException, ExecutionException {
         }
     }
 
+    @Override
+    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+        final long deadlineNanoTime = System.nanoTime() + unit.toNanos(timeout);
+        long nanosToWait;
+        for (; ; ) {
+            if (isCancelled()) {
+                throw new CancellationException();
+            }
+            if (isDone()) {
+                return getResult();
+            }
+            if (Thread.interrupted()) {
+                throw new InterruptedException();
+            }
+
+            nanosToWait = deadlineNanoTime - System.nanoTime();
+            if (nanosToWait < 0) {
+                throw new TimeoutException();
+            }
+
+            synchronized (this) {
```

Comment:
scope of this synchronized block is suspiciously small. the smells like a race-condition to me. usually you want to check your condition while owning the mutex.